File: JavaScript/ExampleTests/common/FandCString.js
import * as tokens from "./tokens.js";

export class FandCString {
  static DNC_STRING = "?DNC?";

  constructor(f = "", c = "", notes = "") {
    this.f = f;
    this.c = c;
    this.notes = notes;
  }

  static fromList(values) {
    const v = Array.from(values);
    return new FandCString(
      v[0] ?? "",
      v[1] ?? "",
      v[2] ?? ""
    );
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text) {
    const parts = tokens.require_(text, 3, "FandC");
    return new FandCString(
      parts[0],
      parts[1],
      parts[2]
    );
  }

  toString() {
    return tokens.token(this.f) + " " + tokens.token(this.c) + " " + tokens.token(this.notes);
  }

  equals(other) {
    if (!(other instanceof FandCString)) return false;
    return (this.f === FandCString.DNC_STRING || other.f === FandCString.DNC_STRING || this.f === other.f)
      && (this.c === FandCString.DNC_STRING || other.c === FandCString.DNC_STRING || this.c === other.c)
      && (this.notes === FandCString.DNC_STRING || other.notes === FandCString.DNC_STRING || this.notes === other.notes);
  }

  static equalLists(a, b) {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]));
  }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/FandCString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class FandCString
    {
        public string f;
        public string c;
        public string notes;

        public FandCString(string f, string c, string notes)
        {
            this.f = f;
            this.c = c;
            this.notes = notes;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static FandCString FromText(string text)
        {
            var parts = Tokens.Require(text, 3, "FandC");
            return new FandCString(parts[0], parts[1], parts[2]);
        }

        public FandCString(string text)
        {
            var parsed = FromText(text);
            this.f = parsed.f;
            this.c = parsed.c;
            this.notes = parsed.notes;
        }

        public FandCTyped ToFandCTyped()
        {
            return new FandCTyped(
                int.Parse(this.f),
                int.Parse(this.c),
                this.notes
            );
        }

        public override string ToString()
        {
            return Tokens.Token(f) + " " + Tokens.Token(c) + " " + Tokens.Token(notes);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not FandCString other) return false;
            return (DNCString == this.f || DNCString == other.f || object.Equals(this.f, other.f))
                && (DNCString == this.c || DNCString == other.c || object.Equals(this.c, other.c))
                && (DNCString == this.notes || DNCString == other.notes || object.Equals(this.notes, other.notes));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.f);
            h.Add(this.c);
            h.Add(this.notes);
            return h.ToHashCode();
        }
    }
}
